tools/testbuild.sh

```diff
@@ -163,83 +138,19 @@ function distclean {
 function configure {
   cd $nuttx/tools || { echo "ERROR: failed to CD to $nuttx/tools"; exit 1; }
   echo "  Configuring..."
-  ./configure.sh $config
-
-  cd $nuttx || { echo "ERROR: failed to CD to $nuttx"; exit 1; }
-
-  if [ "X$host" == "Xlinux" ]; then
-    echo "  Select CONFIG_HOST_LINUX=y"
-
-    kconfig-tweak --file $nuttx/.config --enable CONFIG_HOST_LINUX
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_HOST_WINDOWS
-
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_TOOLCHAIN_WINDOWS
-
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_NATIVE
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_CYGWIN
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_MSYS
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_OTHER
-
-    kconfig-tweak --file $nuttx/.config --enable CONFIG_SIM_X8664_SYSTEMV
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_SIM_X8664_MICROSOFT
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_SIM_M32
-  else
-    echo "  Select CONFIG_HOST_WINDOWS=y"
-    kconfig-tweak --file $nuttx/.config --enable CONFIG_HOST_WINDOWS
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_HOST_LINUX
-
-    if [ "X$wenv" == "Xcygwin" ]; then
-      echo "  Select CONFIG_WINDOWS_CYGWIN=y"
-      kconfig-tweak --file $nuttx/.config --enable CONFIG_WINDOWS_CYGWIN
-      kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_UBUNTU
-      kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_NATIVE
-    else
-      kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_CYGWIN
-      if [ "X$wenv" == "Xubuntu" ]; then
-        echo "  Select CONFIG_WINDOWS_UBUNTU=y"
-        kconfig-tweak --file $nuttx/.config --enable CONFIG_WINDOWS_UBUNTU
-        kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_NATIVE
-      else
-        echo "  Select CONFIG_WINDOWS_NATIVE=y"
-        kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_UBUNTU
-        kconfig-tweak --file $nuttx/.config --enable CONFIG_WINDOWS_NATIVE
-      fi
-    fi
-
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_MSYS
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_WINDOWS_OTHER
-
-    kconfig-tweak --file $nuttx/.config --enable CONFIG_SIM_X8664_MICROSOFT
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_SIM_X8664_SYSTEMV
-    kconfig-tweak --file $nuttx/.config --disable CONFIG_SIM_M32
-  fi
-
-  kconfig-tweak --file $nuttx/.config --disable CONFIG_HOST_MACOS
-  kconfig-tweak --file $nuttx/.config --disable CONFIG_HOST_OTHER
-
-  if [ "X$sizet" == "Xlong" ]; then
```

Comment:
@gregory-nutt here its the code modify CONFIG_CXX_NEWLONG, I remove this option for several reasons:
1.Only testbuild.sh support -s option, other tools(configure.sh/sethost.sh) doesn't. I think these tools should be  consistent with each other.
2.Toolchain is specified in testlist-file, but CONFIG_CXX_NEWLONG must specify through command line, which make it's hard to set these two related options correctly.
How about we change the default value of CXX_NEWLONG base on the compiler config(CONFIG_XXX_TOOLCHAIN_YYY)? so the default value work out of box.

Reply:
You cannot use the compiler to make this decision, because the compiler does not provide the type for size_t.  That is provided by C library or by the C++library.

The same compiler may be built against different libraries to the compiler toolchain selection does not help.

Why don't we use a tristate option:  use default, size_t == unsigned long or size_t == unsigned int.

It cannot be a constant and it cannot be determined by the toolchain (at least not by a "generic" toolchain).

Reply:
> You cannot use the compiler to make this decision, because the compiler does not provide the type for size_t. That is provided by C library or by the C++library.
> 

If so, why we don't define new directly like this:
void *operator new(size_t nbytes)
I can't understand the purpose of CONFIG_CXX_NEWLONG which just make the thing complex:
#ifdef CONFIG_CXX_NEWLONG
void *operator new(unsigned long nbytes)
#else
void *operator new(unsigned int nbytes)
#endif

Reply:
Yes, I think you do not understand how the new and delete operators are provided:  Here are the possibilities:

- The operators could be provided by libs/libxx.  In this case, CONFIG_CXX_NEWLONG determines the argument.
- The operators could be provided by uClibc++ or libc++.  In that case size_t is determined by the architecture specific definitions.  In include/sys/types.h it is set to the type of uinptr_t which varies with architecture
- The operators could be provided by the C++ library binaries built into the toolchain.  These probably come from newlib.  The type of size_t is completely unknown in this case.

I assure you from years of experience with this, you CANNOT use any constant setting.  It must be variable to match all possible cases with all possible tools in the build test.

Reply:
Then there is the other half that also has to be considered:  When the C++ compiler generates binaries, it will use an internal definition for the form of the new and delete operators.

In the failure case here, the build test failes with an undefined reference to void delete(void*, unsigned int);  the new operator is provided by libs/libxx and CONFIG_CXX_NEWLONG is set.  So it provides void delete(void*, unsigned long); and that fails to link in C++ due to the mangled naming.

Reply:
Other compilers may expect void delete(void*, unsigned long);  You cannot know in adance.  Must must be variable or you cannot link consistently.

Reply:
If you just restore the feature so that:

default: Use CONFIG_CXX_NEWLONG from defconfig
-l: Use CONFIG_CXX_NEWLONG =y
-i Use CONFIG_CXX_NEWLONG =n

I will implement that tomorrow.  It is necessary.

Reply:
But it isn't consistent from your explanation:
1.If size_t definition is provided by toolchain(either tool self or c library) and NuttX use that definition directly, it is impossible that what compiler want is different from NuttX provided.
2.Only when size_t is provided by NuttX, it is possible that compiler internal prototype is different from NuttX provided. CONFIG_CXX_NEWLONG can help resolve this mismatch here.
The defintion actually come form nuttx/include/sys/types.h:
/* size_t is used for sizes of memory objects.
 * ssize_t is used for a count of bytes or an error indication.
 *
 * See also definitions of SIZE_MAX et al in limits.h.
 *
 * REVISIT: size_t belongs in stddef.h
 */

#ifdef CONFIG_SMALL_MEMORY

typedef uint16_t     size_t;
typedef int16_t      ssize_t;
typedef uint16_t     rsize_t;

#else /* CONFIG_SMALL_MEMORY */
/* As a general rule, the size of size_t should be the same as the size of
 * uintptr_t: 32-bits on a machine with 32-bit addressing but 64-bits on a
 * machine with 64-bit addressing.
 */

typedef uintptr_t    size_t;
typedef intptr_t     ssize_t;
typedef uintptr_t    rsize_t;

#endif /* CONFIG_SMALL_MEMORY */

Since the code generated by toolchain call many internal functions which is replaced/provided by NuttX(new is just one case). Is it better to  use the basic typedef from toolchain directly? This approach could resolve all potential conflict but without any manual configuration.